File: 0567-permutation-in-string/0567-permutation-in-string.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        int i = 0, m = s1.size(), n = s2.size();
        if(m > n) return false;
        vector<int> cnt1(26, 0), cnt2(26, 0);
        
        for(i=0; i<m; i++){
            cnt1[s1[i]-'a']++;
            cnt2[s2[i]-'a']++;
        }
        
        for(; i<n; i++){
            if(cnt1 == cnt2) return true;
            cnt2[s2[i-m]-'a']--;
            cnt2[s2[i]-'a']++;
        }
        return cnt1 == cnt2 ? true : false;
    }
};

File: 0567-permutation-in-string/0567-permutation-in-string.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        mapp = [0] * 26
        for c in s1:
            mapp[ord(c) - 97] += 1
        i, j, count_chars = 0, 0, len(s1)
        while j < len(s2):
            if mapp[ord(s2[j]) - 97] > 0:   
                count_chars -= 1
            mapp[ord(s2[j]) - 97] -= 1
            j += 1
            if count_chars == 0:
                return True
            if j - i == len(s1):
                if mapp[ord(s2[i]) - 97] >= 0:
                    count_chars += 1
                mapp[ord(s2[i]) - 97] += 1
                i += 1
                
        return False